galley/pkg/config/processing/snapshotter/strategy/debounce.go

```diff
@@ -0,0 +1,185 @@
+	defaultQuiesceDuration = time.Second / 2
+)
+
+// Debounce is a heuristic model for deciding when to publish snapshots. It tries to detect
+// quiesce points for events with a total bounded wait time.
+type Debounce struct {
+	mu sync.Mutex
+
+	maxWaitDuration time.Duration
+	quiesceDuration time.Duration
+
+	changeCh chan struct{}
+	stopCh   chan struct{}
+	doneCh   chan struct{}
+}
+
+var _ Instance = &Debounce{}
+
+// NewDebounceWithDefaults creates a new debounce strategy with default values.
+func NewDebounceWithDefaults() *Debounce {
+	return NewDebounce(defaultMaxWaitDuration, defaultQuiesceDuration)
+}
+
+// NewDebounce creates a new debounce strategy with the given values.
+func NewDebounce(maxWaitDuration, quiesceDuration time.Duration) *Debounce {
+	return &Debounce{
+		maxWaitDuration: maxWaitDuration,
+		quiesceDuration: quiesceDuration,
+		changeCh:        make(chan struct{}, 1),
+	}
+}
+
+// Start implements Instance
+func (d *Debounce) Start(fn OnSnapshotFn) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	if d.stopCh != nil {
+		scope.Debug("Debounce.Start: already started")
+		return
+	}
+	d.stopCh = make(chan struct{})
+	d.doneCh = make(chan struct{})
+
+	// Drain the changeCh, to avoid events from a previous incarnation.
+	drainCh(d.changeCh)
+
+	go d.run(d.stopCh, d.doneCh, fn)
+}
+
+// Stop implements Instance
+func (d *Debounce) Stop() {
+	d.mu.Lock()
+
+	if d.stopCh != nil {
+		scope.Debug("Debounce.Stop: stopping")
+		close(d.stopCh)
+		d.stopCh = nil
+	} else {
+		scope.Debug("Debounce.Stop: already stopped")
+	}
+	d.mu.Unlock()
+
+	<-d.doneCh
+}
+
+func (d *Debounce) run(stopCh, doneCh chan struct{}, fn OnSnapshotFn) {
+	var maxDurationTimer *time.Timer
+	var quiesceTimer *time.Timer
+
+mainloop:
+	for {
+		select {
+		case <-stopCh:
+			break mainloop
+
+		case <-d.changeCh:
+			monitoring.RecordStrategyOnChange()
+			// fallthrough to start the timer.
+		}
+
+		maxDurationTimer = time.NewTimer(d.maxWaitDuration)
+		quiesceTimer = time.NewTimer(d.quiesceDuration)
+
+	loop:
+		for {
+			select {
+			case <-stopCh:
+				break mainloop
+
+			case <-d.changeCh:
+				monitoring.RecordStrategyOnChange()
+
+				quiesceTimer.Stop()
+				drainTimeCh(quiesceTimer.C)
+				quiesceTimer.Reset(d.quiesceDuration)
+				monitoring.RecordOnTimer(false, false, true)
+
+			case <-quiesceTimer.C:
+				monitoring.RecordOnTimer(false, true, false)
+				break loop
+
+			case <-maxDurationTimer.C:
+				monitoring.RecordOnTimer(true, false, false)
+				break loop
+			}
+		}
+
+		quiesceTimer.Stop()
+		drainTimeCh(quiesceTimer.C)
+		maxDurationTimer.Stop()
+		drainTimeCh(maxDurationTimer.C)
+		fn()
+	}
+
+	close(doneCh)
+}
+
+// OnChange implements Instance
+func (d *Debounce) OnChange() {
+	select {
+	case d.changeCh <- struct{}{}:
+	default:
+	}
+}
+
+func drainCh(ch chan struct{}) {
+loop:
+	for {
+		select {
+		case _, ok := <-ch:
+			if !ok {
+				break loop
```

Comment:
Can't you just use "continue" here, rather than needed a loop: label?

Reply:
? I don't think continue works, as we want to exit the loop.

Reply:
I think you meant "break". No it doesn't work. https://play.golang.org/p/ivvmFOVaOwv